File: Java/src/linkedListCode/LinkedList.java
package linkedListCode;

import javax.sound.sampled.Line;

public class LinkedList {
    Node head;

    static class Node{
        int data;
        Node next;

        Node(int d){
            this.data = d;
            next = null;
        }
    }

    static LinkedList insert(LinkedList list, int data){
        Node new_node = new Node(data);

        if(list.head==null){
            list.head = new_node;
        }
        else{
            Node last = list.head;
            while(last.next!=null){
                last = last.next;
            }

            last.next = new_node;
        }

        return list;

    }

    static LinkedList removeNode(LinkedList list, int key){
        Node curr_node = list.head;
        Node prev = null;

        //check if head is the key
        if(curr_node!=null && curr_node.data==key){
            list.head = list.head.next;
            System.out.println("Found and deleted at head!");
            return list;
        }

        //if head is not the key
        while(curr_node!=null && curr_node.data!=key){
            prev = curr_node;
            curr_node = curr_node.next;
        }

        if(curr_node!=null) {
            prev.next = curr_node.next;
            System.out.println("Found and Deleted!");
            return list;
        }

        System.out.println("No key was found!");
        return list;
    }

    static void printList(LinkedList list){
        Node curr_node = list.head;
        if(curr_node == null){
            System.out.println("Empty List");
        }
        else{
            while(curr_node!=null){
                System.out.print(curr_node.data + " ");
                curr_node = curr_node.next;
            }
            System.out.println(" ");
        }
    }

    static void recursivePrint(Node head){
        if(head==null){
            System.out.println(" ");
            return;
        }
        System.out.print(head.data + " ");
        recursivePrint(head.next);
    }

    public static void main(String[] args) {
        LinkedList test_list = new LinkedList();
        insert(test_list, 5);
        insert(test_list, 6);
        insert(test_list, 9);
        insert(test_list, 2);

        printList(test_list);

        removeNode(test_list, 6);

        printList(test_list);
        recursivePrint(test_list.head);

    }
}

File: Python/linkedlist.py
class node:
    def __init__(self, data = None):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def add(self, data):
        new_node = node(data)
        
        # Check if Linked List is empty
        
        if self.head is None:
            self.head = new_node
            return
        
        # If Linked List is not empty
        
        curr = self.head
        while(curr.next != None): # Traverse through list
            curr = curr.next
        curr.next = new_node
        print('Added')
    
    def print_list(self, head=None):
        if head is None:
            head = self.head

        if head is None:
            print("Linked List is Empty")
            return

        curr = head
        result = ""
        while curr:
            result += str(curr.data) + "->"
            curr = curr.next
        result = result.rstrip("->")  # Remove trailing "->"
        print(result)

        
    def length(self):
        if self.head is None:
            print("0")
            return
        
        curr = self.head
        result = 0
        while(curr.next!=None):
            result+=1
            curr = curr.next
        result+=1
        print(result)
        
    def delete(self, key):
        if self.head is None:
            print("Empty List")
            return
        
        if self.head.data is key:
            self.head = self.head.next
            return
        
        curr = self.head
        while curr.next.data!=key:
            curr = curr.next
        
        curr.next = curr.next.next
        print("deleted")
    
    def reverse(self):
        if self.head is None:
            return None

        prev = None
        curr = self.head

        while curr:
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode

        self.head = prev
        self.print_list(self.head)
        return self.head

            

l1 = LinkedList()
l1.add(5)
l1.add(8)
l1.add(10)
l1.add(12)

l1.print_list()
l1.length()

l1.delete(5)
l1.print_list()
l1.length()

l1.reverse()